taichi/ir/control_flow_graph.cpp

```diff
@@ -56,6 +78,163 @@ bool CFGNode::erase_entire_node() {
   return true;
 }
 
+void CFGNode::reaching_definition_analysis() {
+  // Calculate reach_gen and reach_kill.
+  reach_gen.clear();
+  reach_kill.clear();
+  for (int i = end_location - 1; i >= begin_location; i--) {
+    // loop in reversed order
+    auto stmt = block->statements[i].get();
+    auto data_source_ptr = irpass::analysis::get_data_source_pointer(stmt);
+    if (data_source_ptr) {
+      // stmt provides a data source
+      // TODO: If stmt is a GlobalPtrStmt or a GlobalTemporaryStmt, we may lose
+      // optimization opportunities in this case (we could replace $4 with b):
+      // $1: global ptr a
+      // $2: global store [$1 <- b]
+      // $3(stmt): global ptr a
+      // $4: global load $3
+      if (!reach_kill_variable(data_source_ptr)) {
+        reach_gen.insert(stmt);
+        reach_kill.insert(data_source_ptr);
+      }
+    }
+  }
+}
+
+bool CFGNode::reach_kill_variable(Stmt *var) const {
+  // Does this node (definitely) kill a definition of var?
+  // return reach_kill.find(var) != reach_kill.end();
```

Comment:
If we return this directly, the running time of `reaching_definition_analysis` here https://github.com/taichi-dev/taichi/issues/926#issuecomment-645092561 will decrease from `45.799 ms` to `14.094 ms`.